CSUB-518: Pt. 1 - introduced set_gate_contract

zacharyfrederick opened the pull request:
This PR makes several important changes:
1. It introduces the `set_gate_contract` extrinsic
2. It renames the old `GCreContract` type to `DeployedContract`.

## Rationale:
1. In order to facilitate testing the swap token mechanics locally and on non-gluwa controlled test networks we allow the administrator of the blockchain to set the on-chain storage values for the GATE and GCRE contracts (the address that the contracts are deployed to, necessary for verifying function calls using the tx hash). This particular PR adds the extrinsic for the GATE contract as well as its associated on-chain storage. The extrinsic for setting the GCRE contract has been implemented previously. 
2. We abstracted the GCreContract to represent any deployed contract on an ethereum like blockchain. This required updating the integration tests for the previous GCRE->CTC functionality. The current `request_collect_coins` implementation is built entirely around the GCRE contract. This change will be more relevant when the `request_collect_coins_v2` extrinsic is introduced. 


## Tests for `set_burn_gate_contract`
	- [x] When not set the storage value should return the default goerli contract
	- [x] The extrinsic can only be set by a root user
	- [x] The extrinsic can be set and the storage value is persisted

Commit:
Introduce the set_gate_contract() extrinsic

in preparation for request_collect_coins_v2

atodorov commented on pallets/creditcoin/src/tests.rs:
```diff
@@ -3236,3 +3237,47 @@ fn register_address_v2_should_error_with_unsupported_blockchain() {
+}
+
+#[test]
+fn set_burn_gate_contract_fails_with_non_root() {
```
```suggestion
fn set_gate_contract_should_fail_when_not_signed_by_root() {
```

Naming:
- prefix with the actual extrinsic name to make tests easier to find
- test name is more explicit

atodorov commented on pallets/creditcoin/src/tests.rs:
```diff
@@ -3236,3 +3237,47 @@ fn register_address_v2_should_error_with_unsupported_blockchain() {
+}
+
+#[test]
+fn set_burn_gate_contract_passes_and_storage_is_updated() {
```
```suggestion
fn set_gate_contract_passes_and_storage_is_updated() {
```

- prefix with the actual extrinsic name to make tests easier to find

atodorov commented on pallets/creditcoin/src/tests.rs:
```diff
@@ -3236,3 +3237,47 @@ fn register_address_v2_should_error_with_unsupported_blockchain() {
+
+		let gate_contract =
+			DeployedContract { address: fake_address, chain: Blockchain::Luniverse };
+
```
stricter assertions

```suggestion

		assert_ne!(gate_contract, DeployedContract::default());
		assert_eq!(Creditcoin::gate_contract(), DeployedContract::default());
```

atodorov commented on pallets/creditcoin/src/tests.rs:
```diff
@@ -3236,3 +3237,47 @@ fn register_address_v2_should_error_with_unsupported_blockchain() {
+use hex_literal::hex;
+
+#[test]
+fn set_burn_gate_contract_should_return_default_goerli_contract_when_not_set() {
```
```suggestion
fn gate_contract_storage_should_return_default_goerli_contract_when_not_set() {
```

because this test has nothing to do with the `set_gate_contract()` extrinsic.

atodorov commented on pallets/creditcoin/src/tests.rs:
```diff
@@ -3236,3 +3237,47 @@ fn register_address_v2_should_error_with_unsupported_blockchain() {
+			sp_core::H160(hex!("a3EE21C306A700E682AbCdfe9BaA6A08F3820419"))
+		);
+
+		assert_eq!(contract.chain, Blockchain::Ethereum);
```
Why not just 

```
	assert_eq!(contract, DeployedContract::default());
```

atodorov commented on pallets/creditcoin/src/tests.rs:
```diff
@@ -3236,3 +3237,47 @@ fn register_address_v2_should_error_with_unsupported_blockchain() {
 		);
 	});
 }
```
Inside `fn exercise_weightinfo_functions()` add this snippet

```
	let result = super::weights::WeightInfo::<Test>::set_gate_contract();
	assert!(result.ref_time() > 0);
```

atodorov reviewed (changes requested):
Needs some adjustments for tests, otherwise looks good.

Note: I will merge the suggestions & squash some of the commits in the interest of fast turnaround time.

atodorov committed:
Update test names and add more assertions

- fix initial WeightInfo to be greater than zero before benchmarks
  have the chance to update this appropriately

Commit:
Auto-update creditcoin-js type definitions
